### ISSUE TITLE
10

### ISSUE THREAD

**Issue by dingding800** [dingding800/HomeAutomation#11]
# Install required libraries (uncomment the next line if running for the first time)
# !pip install pandas scikit-learn fuzzywuzzy[speedup]

import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.neighbors import NearestNeighbors
from fuzzywuzzy import process

# Define standard occupation titles
standard_occupations = [
    "Software Engineer",
    "Doctor",
    "Lawyer",
    "Accountant",
    "Data Scientist",
    "Real Estate Agent",
    "Consultant",
    "Teacher",
    "Nurse",
    "Construction Worker",
    "Taxi Driver",
    "Waiter",
    "Chef",
    "Bank Teller"
]

# Fuzzy matching approach
def fuzzy_standardize(occupation, choices, threshold=80):
    match, score = process.extractOne(occupation, choices)
    return match if score >= threshold else "Other"

# TF-IDF + Nearest Neighbors approach
class OccupationStandardizer:
    def __init__(self, standard_titles):
        self.vectorizer = TfidfVectorizer(analyzer='char_wb', ngram_range=(2, 4))
        self.standard_titles = standard_titles
        self.nn = NearestNeighbors(n_neighbors=1, metric='cosine')
        self._fit()

    def _fit(self):
        X = self.vectorizer.fit_transform(self.standard_titles)
        self.nn.fit(X)

    def standardize(self, raw_occupation):
        X_query = self.vectorizer.transform([raw_occupation])
        distance, index = self.nn.kneighbors(X_query)
        if distance[0][0] < 0.4:
            return self.standard_titles[index[0][0]]
        else:
            return "Other"

# Sample raw occupation data
raw_data = pd.DataFrame({
    "raw_occupation": [
        "sw eng", "dev", "cook", "real estate broker", "lawyer", "medical doctor",
        "Data Eng", "construction labour", "wait staff", "school teacher"
    ]
})

# Option A: Use fuzzy matching
raw_data["standard_occupation_fuzzy"] = raw_data["raw_occupation"].apply(
    lambda x: fuzzy_standardize(x, standard_occupations)
)

# Option B: Use TF-IDF + Nearest Neighbors
standardizer = OccupationStandardizer(standard_occupations)
raw_data["standard_occupation_ml"] = raw_data["raw_occupation"].apply(
    standardizer.standardize
)

# Print results
print(raw_data)
